nwbwidgets/ophys/single_plane.py

```diff
@@ -0,0 +1,202 @@
+
+        self.setup_observers()
+
+        # Setup layout of Canvas relative to Controllers
+        self.children = [self.Canvas, self.Controller]
+
+    @lru_cache  # default size of 128 items ought to be enough to create a 1GB cache on large images
+    def _cache_data_read(self, dataset: h5py.Dataset, frame_index: int, plane_index: int) -> np.ndarray:
+        return dataset[frame_index, :, :, plane_index].T
+
+    def update_contrast_range(self):
+        """
+        If either of the frame or plane sliders are changed, be sure to update the valid range of the manual contrast.
+
+        Applies even if current hidden, in case user wants to enable it.
+        """
+        self.Controller.manual_contrast_slider.max = np.max(self.data)
+        self.Controller.manual_contrast_slider.min = np.min(self.data)
+        self.Controller.manual_contrast_slider.value = (
+            max(self.Controller.manual_contrast_slider.value[0], self.Controller.manual_contrast_slider.min),
+            min(self.Controller.manual_contrast_slider.value[1], self.Controller.manual_contrast_slider.max),
+        )
+
+    def update_data(self, frame_index: Optional[int] = None):
+        frame_index = frame_index or self.Controller.frame_slider.value
+
+        self.data = self._cache_data_read(dataset=self.two_photon_series.data, frame_index=frame_index)
+
+        self.data = self.two_photon_series.data[frame_index, :, :]
+
+        if self.Controller.contrast_type_toggle.value == "Manual":
+            self.update_contrast_range()
+
+    def setup_data(self, max_mb_treshold: float = 20.0):
+        """
+        Start by loading only a single frame of a single plane.
+
+        If the image size relative to data type is too large, relative to max_mb_treshold (indicating the load
+        operation for initial setup would take a noticeable amount of time), then sample the image with a `by`.
+
+        Note this may not actually provide a speedup when streaming; need to think of way around that. Maybe set
+        a global flag for if streaming mode is enabled on the file, and if so make full use of data within contiguous
+        HDF5 chunks?
+        """
+        itemsize = self.two_photon_series.data.dtype.itemsize
+        nbytes_per_image = math.prod(self.two_photon_series.data.shape) * itemsize
+        if nbytes_per_image <= max_mb_treshold:
+            self.update_data(frame_index=0)
+        else:
+            # TOD: Figure out formula for calculating by in one-shot
+            by_width = 2
+            by_height = 2
+            self.data = self.two_photon_series.data[0, ::by_width, ::by_height]
+
+    def get_rotation(self) -> int:
+        """The rotation attribute of the SinglePlaneDataController cannot be attached in a modifiable state."""
+        if not hasattr(self, "Controller"):  # First time this is called
+            return 0
+        return self.Controller.components[self.data_controller_name].components["RotationController"].rotation
```

Comment:
This is the only time a reference to a nested child is necessary (due to attribute mutability relative to the button observer events)

Technically I could override those observers and remap it all to an attribute specific to these classes, but that ends up taking many more lines of code than this simple router

That said, I built the `MutliController` to at least have dictionary mapping capability, so you can navigate it entirely based on name references to the components (much easier than children indexes)